### ISSUE TITLE
[Python] Cannot run `goss` using custom objective function `fobj`

### ISSUE THREAD

**Issue by PastelBelem8** [microsoft/LightGBM#3243]
## How you are using LightGBM?

LightGBM component: Python package

## Environment info

Operating System: Ubuntu 18.04.4 LTS

CPU/GPU model: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz (x86_64) / Intel Corporation UHD Graphics 620

C++ compiler version: gcc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0

CMake version: 3.10.2

Java version: openjdk version "1.8.0_252"
OpenJDK Runtime Environment (build 1.8.0_252-8u252-b09-1~18.04-b09)
OpenJDK 64-Bit Server VM (build 25.252-b09, mixed mode)

Python version: 3.7.7

LightGBM version or commit hash: [f257636d](https://github.com/microsoft/LightGBM/tree/f257636d01dab724ed7a800cde20e68d5ebcc867) (latest commit).

## Error message and / or logs

```bash
[LightGBM] [Warning] Using self-defined objective function
[LightGBM] [Warning] Auto-choosing row-wise multi-threading, the overhead of testing was 0.000101 seconds.
You can set `force_row_wise=true` to remove the overhead.
And if memory is not enough, you can set `force_col_wise=true`.
[LightGBM] [Info] Total Bins 2550
[LightGBM] [Info] Number of data points in the train set: 10000, number of used features: 10
[LightGBM] [Info] Using GOSS
[LightGBM] [Warning] Using self-defined objective function
[LightGBM] [Info] Using GOSS
[LightGBM] [Info] Using GOSS
[1]    30327 segmentation fault (core dumped)  python -m goss_fobj
```

## Reproducible example(s)

```python
import numpy as np
import sklearn.datasets as sd
import lightgbm as lgb

# Auxiliar Functions
def sigmoid(x: np.ndarray) -> np.ndarray:
    """A numerically stable version of the logistic sigmoid function."""
    pos_mask = (x >= 0)
    neg_mask = (x < 0)
    z = np.zeros_like(x)
    z[pos_mask] = np.exp(-x[pos_mask])
    z[neg_mask] = np.exp(x[neg_mask])
    top = np.ones_like(x)
    top[neg_mask] = z[neg_mask]

    return top / (1 + z)

def x_entropy_udf(preds: np.ndarray, d: lgb.Dataset) -> tuple:
    labels = d.get_label()

    preds = sigmoid(preds)
    grads = preds - labels
    hesss = preds * (1.0 - preds)
    return grads, hesss


# Constants
SEED = 101
N_ROUNDS = 50
N_SAMPLES = 10 * 1000


X, y = sd.make_hastie_10_2(N_SAMPLES, random_state=SEED)
data = lgb.Dataset(X, y)
params = {
    'objective': None,
    'boosting': 'goss', 
    'seed': SEED,
}
lgb.train(params, data, num_boost_round=N_ROUNDS, fobj=x_entropy_udf)
```
## Steps to reproduce

1. Create conda environment with `python=3.7.7`
2. Activate environment and `python -m pip install setuptools wheel numpy scipy scikit-learn -U`
3. Download code from commit [f257636d](https://github.com/microsoft/LightGBM/tree/f257636d01dab724ed7a800cde20e68d5ebcc867), change directory to `LightGBM/python_package` and run `python -m setup install`
4. Move to home directory and run code above.

## Notes 

I'm able to run the code above for `N_ROUNDS=10`, however, raising this value leads to a segmentation fault. Neither the previous issues nor the posts in StackOverflow ([segmentation-fault-when-defining-a-custom-loss-in-lightgbm](https://stackoverflow.com/questions/60120979/segmentation-fault-when-defining-a-custom-loss-in-lightgbm)) have an answer to this error. 

As suggested in issue [#2156](https://github.com/microsoft/LightGBM/issues/2156),  I even tried running the [`advanced_example.py`](https://github.com/microsoft/LightGBM/blob/74ce2cfe9418b0def315c5e17f6244eefac64a60/examples/python-guide/advanced_example.py#L156-L162) but instead passed the parameters defined above. It still crashes.

**Comment by guolinke** [comment]
@PastelBelem8 Did this happen only when using the customed objective? 
@shiyu1994 can you help to check this issue?

**Comment by PastelBelem8** [comment]
Yes @guolinke

**Comment by shiyu1994** [comment]
@guolinke ok.

**Comment by guolinke** [comment]
@PastelBelem8 
Does
```
params = {
    'objective': 'binary', 
    'boosting': 'goss',
    'seed': SEED,
}
lgb.train(params, data, num_boost_round=N_ROUNDS)
```
also work?

**Comment by PastelBelem8** [comment]
@guolinke  yes, it does!

**Comment by PastelBelem8** [comment]
@guolinke any tip that might get me started on debugging potential reasons for this error? I would like to contribute if possible :)

**Comment by guolinke** [comment]
Thanks @PastelBelem8 , it will be great!
firstly, you can try `force_col_wise=true`.
if there is not problem, I think 
I think it is caused by the subset used in bagging:
https://github.com/microsoft/LightGBM/blob/952e23a0093e68cde6453c7ac2f4a818b5892c81/src/boosting/goss.hpp#L145-L154

if with  `force_col_wise=true`, it still failed. I think the problem is caused by bagging:
https://github.com/microsoft/LightGBM/blob/952e23a0093e68cde6453c7ac2f4a818b5892c81/src/boosting/goss.hpp#L134-L142

**Comment by PastelBelem8** [comment]
@guolinke for this particular dataset, setting `force_col_wise=true' seems to solve the problem. Thanks for the help.

**Comment by StrikerRUS** [comment]
@PastelBelem8 Thanks a lot for the updates!

However, I don't think that a segfault is something that should happen during training with custom objective. So I'd better reopen the issue.
